[WPE] Add gamepad support

ceyusa opened the pull request:
#### ab0e24752e11bbdb584662dcd5369ce17f4da1de
<pre>
[WPE] Add gamepad support
<a href="https://bugs.webkit.org/show_bug.cgi?id=230630">https://bugs.webkit.org/show_bug.cgi?id=230630</a>

With gamepad support in libwpe[1] is possible to communicate gamepads events between the
application (cog, for example) and WPEWebKit. This patch are the changes required to attend that
communication and enable the gamepad support in WPE port.

1. <a href="https://github.com/WebPlatformForEmbedded/libwpe/pull/88">https://github.com/WebPlatformForEmbedded/libwpe/pull/88</a>

Reviewed by Adrian Perez de Castro.

* LayoutTests/platform/wpe/TestExpectations: Enable skipped tests.
* Source/WebCore/PlatformWPE.cmake: Add headers if enabled.
* Source/WebCore/SourcesWPE.txt: Add source files.
* Source/WebCore/platform/gamepad/wpe/WPEGamepad.cpp: Added.
(WebCore::WPEGamepad::WPEGamepad):
(WebCore::WPEGamepad::~WPEGamepad):
(WebCore::WPEGamepad::buttonPressedOrReleased):
(WebCore::WPEGamepad::absoluteAxisChanged):
* Source/WebCore/platform/gamepad/wpe/WPEGamepad.h: Added.
* Source/WebCore/platform/gamepad/wpe/WPEGamepadProvider.cpp: Added.
(WebCore::WPEGamepadProvider::singleton):
(WebCore::WPEGamepadProvider::WPEGamepadProvider):
(WebCore::WPEGamepadProvider::~WPEGamepadProvider):
(WebCore::WPEGamepadProvider::startMonitoringGamepads):
(WebCore::WPEGamepadProvider::stopMonitoringGamepads):
(WebCore::WPEGamepadProvider::gamepadConnected):
(WebCore::WPEGamepadProvider::gamepadDisconnected):
(WebCore::WPEGamepadProvider::indexForNewlyConnectedDevice):
(WebCore::WPEGamepadProvider::removeGamepadForId):
(WebCore::WPEGamepadProvider::initialGamepadsConnectedTimerFired):
(WebCore::WPEGamepadProvider::inputNotificationTimerFired):
(WebCore::WPEGamepadProvider::scheduleInputNotification):
(WebCore::WPEGamepadProvider::inputView):
* Source/WebCore/platform/gamepad/wpe/WPEGamepadProvider.h: Added.
* Source/WebKit/SourcesWPE.txt: Add source files.
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp: Reports the view where gamepad currently operates.
(WKWPE::m_backend):
(WKWPE::View::~View):
(WKWPE::View::setViewState):
(WKWPE::View::platformWebPageProxyForGamepadInput):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:
* Source/WebKit/UIProcess/Gamepad/UIGamepadProvider.cpp:
* Source/WebKit/UIProcess/Gamepad/wpe/UIGamepadProviderWPE.cpp: Added.
(WebKit::UIGamepadProvider::platformSetDefaultGamepadProvider):
(WebKit::UIGamepadProvider::platformWebPageProxyForGamepadInput):
(WebKit::UIGamepadProvider::platformStopMonitoringInput):
(WebKit::UIGamepadProvider::platformStartMonitoringInput):
* Source/cmake/OptionsWPE.cmake:

Canonical link: <a href="https://commits.webkit.org/253720@main">https://commits.webkit.org/253720@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/90e79f0c207f737824efee8436e574b989610645

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86800 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30859 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17697 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95633 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149384 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90781 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29243 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25619 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78962 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90868 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92416 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23616 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73690 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23655 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78614 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/78925 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66660 "Found 1 new API test failure: /WebKit2Gtk/TestWebKitWebView:/webkit/WebKitWebView/is-web-process-responsive") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78721 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27005 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12779 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72371 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26927 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13793 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25829 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2638 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28611 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36659 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75155 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28555 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33072 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16630 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -77,7 +77,11 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_SHAREABLE_RESOURCE PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEB_RTC PRIVATE ${ENABLE_EXPERIMENTAL_FEATURES})
-WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEBXR PRIVATE ${ENABLE_EXPERIMENTAL_FEATURES})
+
+if (${ENABLE_EXPERIMENTAL_FEATURES} AND "${WPE_VERSION}" VERSION_GREATER_EQUAL "1.13.1")
```
I think you can write this more simply as:

```suggestion
if (ENABLE_EXPERIMENTAL_FEATURES AND ${WPE_VERSION} VERSION_GREATER_EQUAL 1.13.1)
```

That said, gating it behind ENABLE_EXPERIMENTAL_FEATURES is a big hammer to ensure that this functionality is not actually available to users. It's intended for half-baked features that users should not be able to enable without recompiling WebKit on their own. Is that really desired here? If it's more or less ready for users, sans bugs, then it's probably best to just enable it.

aperezdc replied:
Building the gamepad support only depends on a recent-enough libwpe version. We should enable it by default, and packagers which have older versions of libwpe can always configure their builds with `-DENABLE_GAMEPAD=OFF`.

ceyusa replied:
Sorry for the delay! I have updated it accordingly with your comments. Thanks @mcatanzaro

aperezdc commented on Source/WebCore/platform/gamepad/wpe/WPEGamepadProvider.cpp:
```diff
@@ -0,0 +1,222 @@
+    if (index != notFound)
+        m_gamepadVector[index] = nullptr;
+
+    return removedGamepad;
```
This function should reset `m_lastActiveGamepad` to null if the gamepad being removed is the last active one.

aperezdc commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -84,6 +84,10 @@ WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEB_RTC PRIVATE ${ENABLE_EXPERIMENTAL_FE
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEBGL2 PRIVATE ON)
 WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_WEBXR PRIVATE ${ENABLE_EXPERIMENTAL_FEATURES})
 
+if (WPE_VERSION VERSION_GREATER_EQUAL 1.13.1)
```
We should check for `1.13.90` or newer here, which is the version that has the most recent gamepad API which is the most likely to make it to the next stable libwpe.

ceyusa replied:
done

aperezdc commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -199,6 +203,10 @@ if (ENABLE_ACCESSIBILITY)
     endif ()
 endif ()
 
+if (ENABLE_GAMEPAD AND (NOT (WPE_VERSION VERSION_GREATER_EQUAL 1.13.3)))
```
Same here, let's check for `1.13.90`.

ceyusa replied:
done

aperezdc reviewed (changes requested):
This needs a couple of touch-ups before landing, but it is almost ready.

aperezdc reviewed (approved):
r=me

ceyusa committed:
[WPE] Add gamepad support
https://bugs.webkit.org/show_bug.cgi?id=230630

With gamepad support in libwpe[1] is possible to communicate gamepads events between the
application (cog, for example) and WPEWebKit. This patch are the changes required to attend that
communication and enable the gamepad support in WPE port.

1. https://github.com/WebPlatformForEmbedded/libwpe/pull/88

Reviewed by Adrian Perez de Castro.

* LayoutTests/platform/wpe/TestExpectations: Enable skipped tests.
* Source/WebCore/PlatformWPE.cmake: Add headers if enabled.
* Source/WebCore/SourcesWPE.txt: Add source files.
* Source/WebCore/platform/gamepad/wpe/WPEGamepad.cpp: Added.
(WebCore::WPEGamepad::WPEGamepad):
(WebCore::WPEGamepad::~WPEGamepad):
(WebCore::WPEGamepad::buttonPressedOrReleased):
(WebCore::WPEGamepad::absoluteAxisChanged):
* Source/WebCore/platform/gamepad/wpe/WPEGamepad.h: Added.
* Source/WebCore/platform/gamepad/wpe/WPEGamepadProvider.cpp: Added.
(WebCore::WPEGamepadProvider::singleton):
(WebCore::WPEGamepadProvider::WPEGamepadProvider):
(WebCore::WPEGamepadProvider::~WPEGamepadProvider):
(WebCore::WPEGamepadProvider::startMonitoringGamepads):
(WebCore::WPEGamepadProvider::stopMonitoringGamepads):
(WebCore::WPEGamepadProvider::gamepadConnected):
(WebCore::WPEGamepadProvider::gamepadDisconnected):
(WebCore::WPEGamepadProvider::indexForNewlyConnectedDevice):
(WebCore::WPEGamepadProvider::removeGamepadForId):
(WebCore::WPEGamepadProvider::initialGamepadsConnectedTimerFired):
(WebCore::WPEGamepadProvider::inputNotificationTimerFired):
(WebCore::WPEGamepadProvider::scheduleInputNotification):
(WebCore::WPEGamepadProvider::inputView):
* Source/WebCore/platform/gamepad/wpe/WPEGamepadProvider.h: Added.
* Source/WebKit/SourcesWPE.txt: Add source files.
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp: Reports the view where gamepad currently operates.
(WKWPE::m_backend):
(WKWPE::View::~View):
(WKWPE::View::setViewState):
(WKWPE::View::platformWebPageProxyForGamepadInput):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:
* Source/WebKit/UIProcess/Gamepad/UIGamepadProvider.cpp:
* Source/WebKit/UIProcess/Gamepad/wpe/UIGamepadProviderWPE.cpp: Added.
(WebKit::UIGamepadProvider::platformSetDefaultGamepadProvider):
(WebKit::UIGamepadProvider::platformWebPageProxyForGamepadInput):
(WebKit::UIGamepadProvider::platformStopMonitoringInput):
(WebKit::UIGamepadProvider::platformStartMonitoringInput):
* Source/cmake/OptionsWPE.cmake:

Canonical link: https://commits.webkit.org/253720@main
